chapter_15_advanced_forms.asciidoc

```diff
@@ -908,18 +764,78 @@ NOTE: Personal opinion here: I could have used `super`, but I prefer not to use
     Anything else is too error-prone, and I find it ugly besides. YMMV.
 
 
-And we're there!  All the unit tests pass:
+Let's run the tests!  All the unit tests pass:
 
 [subs="specialcharacters,macros"]
 ----
 $ pass:quotes[*python manage.py test lists*]
 [...]
-Ran 34 tests in 0.082s
+Ran 30 tests in 0.082s
 
 OK
 ----
 
-And so does our FT for validation:
+But we still have something to do about our FTs:
+[subs="specialcharacters,macros"]
+----
+$ pass:quotes[*python manage.py test functional_tests.test_list_item_validation*]
+[...]
+FAIL: test_cannot_add_duplicate_items (functional_tests.test_list_item_validation.ItemValidationTest.test_cannot_add_duplicate_items)
+----------------------------------------------------------------------
+[...]
+AssertionError: '' != "You've already got this in your list"
++ You've already got this in your list
+----
+
+The error message isn't being displayed because we are not using the Bootstrap
+classes. Although it would have been nice to minimise hand-written html and
+utilise Django instead, it seems like we need to bring back our custom
+`<input>` and add a few attributes manually:
+
+[role="sourcecode"]
+.src/lists/templates/base.html
+====
+[source,diff]
+----
+@@ -13,13 -14,26 +14,26 @@@
+  <div class="col-lg-6 text-center">
+      <h1 class="display-1 mb-4">{% block header_text %}{% endblock %}</h1>
+-     <form method="POST" action="{% block form_action %}{% endblock %}">
+-         {{ form.text }}
++     <form
++         class="was-validated"  <1>
++         method="POST"
++         action="{% block form_action %}{% endblock %}"
++     >
+          {% csrf_token %}
++         <input
++             id="id_text"
++             name="text"
++             class="form-control form-control-lg {% if form.errors %}is-invalid{% endif %}"  <2>
++             placeholder="Enter a to-do item"
++             value="{{ form.text.value }}"
++             aria-describedby="id_text_feedback"
++             required
++         />
+          {% if form.errors %}
+-             <div class="invalid-feedback">{{ form.errors.text }}</div>
++             <div class="invalid-feedback">{{ form.errors.text.0 }}</div>  <3>
+          {% endif %}
+      </form>
+  </div>
+----
+====
+
+<1> We give `.was-validated` to the `<form>`
+<2> We provide an `<input>` and the most important custom setting will be its
+    `class`. As you can see, we can use conditionals even for providing
+    additional `class` -es.
+<3> If you try to just simply display `form.errors.text` you'll see it ends up
+    as a list item in an unordered list. To avoid that, let's just
+    select `form.errors.text.0`.
+
```

Comment:
```suggestion

NOTE:  Another fliip-flop! We spent most of the last chapter switching from handcrafted html
  to having our form autogenerated by django, and now we're switching back.
  It's a little frustrating, and I could have gone back and changed the book's text to avoid the back+forth,
  but I prefer to show software development as it really is.
  We often try things out and end up changing our minds.
  Particularly with frameworks like Django,
  you can find yourself taking advantage of auto-generated shortcuts for as long as they work,
  but at some points you meet the limits of what the framework designers have anticipated,
  and it's time to go back to doing the work yourself.
  It doesn't mean you should always reinvent the wheel!
```